blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,376 @@
+    ## Consequently this assumes that message
+    ## is the output of a fixed size hash function.
+    signature: Signature
+    pubkey: PublicKey
+    message: array[HashLen, byte]
+
+  BatchedBLSVerifier*[HashLen: static int] = object
+    ## A type to batch BLS multi signatures (aggregated or individual)
+    ## verification using multiple cores if compiled with OpenMP
+    sets: seq[SignatureSet[HashLen]]
+
+    # Per-batch contexts for multithreaded batch verification
+    batchContexts: seq[ContextMultiAggregateVerify[DST]]
+
+func init*(T: type BatchedBLSVerifier): T {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  # Impl: A BatchedBLSVerifier still MUST be zero-init
+  # otherwise the sequences fields
+  # like capacity and reserve memory will be wrong.
+  result.sets.setLen(0)
+  result.batchContexts.setLen(0)
+
+func clear*(batcher: var BatchedBLSVerifier) {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  batcher.sets.setLen(0)
+  batcher.batchContexts.setLen(0)
+
+func incl*[HashLen: static int](
+       batcher: var BatchedBLSVerifier[HashLen],
+       public_key: PublicKey,
+       message: array[HashLen, byte],
+       signature: Signature
+     ): bool {.inline.} =
+  ## Include a (public key, message, signature) triplet
+  ## to the batch for verification.
+  ##
+  ## Always return true
+  ##
+  ## The proof-of-possession MUST be verified before calling this function.
+  ## It is recommended to use the overload that accepts a proof-of-possession
+  ## to enforce correct usage.
+  batcher.sets.add SignatureSet[HashLen](
+    signature: signature,
+    pubkey: public_key,
+    message: message
+  )
+  return true
+
+func incl*[HashLen: static int](
+       batcher: var BatchedBLSVerifier[HashLen],
+       public_keys: openarray[PublicKey],
+       message: array[HashLen, byte],
+       signature: Signature
+     ): bool {.inline.} =
+  ## Include a (array of public keys, message, signature) triplet
+  ## to the batch for verification.
+  ##
+  ## All public keys sign the same message
+  ## and signature is their aggregated signature
+  ##
+  ## Returns false if no public keys are passed
+  ## Returns true otherwise
+  ##
+  ## The proof-of-possession MUST be verified before calling this function.
+  ## It is recommended to use the overload that accepts a proof-of-possession
+  ## to enforce correct usage.
+  if publicKeys.len == 0:
+    return false
+
+  var aggAffine{.noInit.}: PublicKey
+  if not aggAffine.aggregateAll(publicKeys):
+    return false
+
+  batcher.sets.add SignatureSet(
+    signature: signature,
+    pubkey: aggAffine,
+    message: message
+  )
+
+# Serial Batch Verifier
+# ----------------------------------------------------------------------
+
+func batchVerifySerial*(
+       batcher: BatchedBLSVerifier,
+       secureRandomBytes: array[32, byte]
+     ): bool =
+  ## Single-threaded batch verification
+  if batcher.sets.len == 0:
+    return false
```

Comment:
return `true`? logically, no verifications failed (ie this is an `all` style operation)

Reply:
Spec requirement: https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-04#section-3.1.1
` Precondition: n >= 1, otherwise return INVALID.`

```Nim

3.1.1.  AggregateVerify

   This function first ensures that all messages are distinct, and then
   invokes CoreAggregateVerify.

   result = AggregateVerify((PK_1, ..., PK_n),
                            (message_1, ..., message_n),
                            signature)

   Inputs:
   - PK_1, ..., PK_n, public keys in the format output by SkToPk.
   - message_1, ..., message_n, octet strings.
   - signature, an octet string output by Aggregate.

   Outputs:
   - result, either VALID or INVALID.

   Precondition: n >= 1, otherwise return INVALID.

   Procedure:
   1. If any two input messages are equal, return INVALID.
   2. return CoreAggregateVerify((PK_1, ..., PK_n),
                                 (message_1, ..., message_n),
                                 signature)
```